Recast/Include/RecastAlloc.h

```diff
@@ -58,64 +59,252 @@ void* rcAlloc(size_t size, rcAllocHint hint);
-	///  @param[in]		item	The new value.
-	void push(int item) { resize(m_size+1); m_data[m_size-1] = item; }
+	rcSizeType new_cap = m_size ? 2*m_size : 1;
+	T* data = allocate_and_copy(new_cap);
+	// construct between allocate and destroy+free in case value is
+	// in this vector.
+	construct(data + m_size, value);
+	destroy_range(0, m_size);
+	m_size++;
+	m_cap = new_cap;
+	rcFree(m_data);
+	m_data = data;
+}
+template <typename T, rcAllocHint H>
+void rcVectorBase<T, H>::resize_impl(rcSizeType size, const T* value) {
+	if (size < m_size) {
+		destroy_range(size, m_size);
+		m_size = size;
+	} else if (size > m_size) {
+		T* new_data = allocate_and_copy(size);
+		// We defer deconstructing/freeing old data until after constructing
+		// new elements in case "value" is there.
+		if (value) {
+			construct_range(new_data + m_size, new_data + size, *value);
+		} else {
+			construct_range(new_data + m_size, new_data + size);
+		}
+		destroy_range(0, m_size);
+		rcFree(m_data);
+		m_data = new_data;
+		m_cap = size;
+		m_size = size;
+	}
+}
+template <typename T, rcAllocHint H>
+void rcVectorBase<T, H>::swap(rcVectorBase<T, H>& other) {
+	// TODO: Reorganize headers so we can use rcSwap here.
+	rcSizeType tmp_cap = other.m_cap;
+	rcSizeType tmp_size = other.m_size;
+	T* tmp_data = other.m_data;
 
-	/// Returns the value at the end of the array and reduces the size by one.
-	///  @return The value at the end of the array.
-	int pop()
-	{
-		if (m_size > 0)
-			m_size--;
-		
-		return m_data[m_size];
+	other.m_cap = m_cap;
+	other.m_size = m_size;
+	other.m_data = m_data;
+
+	m_cap = tmp_cap;
+	m_size = tmp_size;
+	m_data = tmp_data;
+}
+// static
+template <typename T, rcAllocHint H>
+void rcVectorBase<T, H>::construct_range(T* begin, T* end) {
+	for (T* p = begin; p < end; p++) {
+		construct(p);
+	}
+}
+// static
+template <typename T, rcAllocHint H>
+void rcVectorBase<T, H>::construct_range(T* begin, T* end, const T& value) {
+	for (T* p = begin; p < end; p++) {
+		construct(p, value);
+	}
+}
+// static
+template <typename T, rcAllocHint H>
+void rcVectorBase<T, H>::copy_range(T* dst, const T* begin, const T* end) {
+	for (rcSizeType i = 0 ; i < end - begin; i++) {
+		construct(dst + i, begin[i]);
 	}
+}
+template <typename T, rcAllocHint H>
+void rcVectorBase<T, H>::destroy_range(rcSizeType begin, rcSizeType end) {
+	for (rcSizeType i = begin; i < end; i++) {
+		m_data[i].~T();
+	}
+}
+template <typename T, rcAllocHint H>
+rcVectorBase<T, H>& rcVectorBase<T, H>::operator=(const rcVectorBase<T, H>& other) {
+	if (this != &other) {
+		assign(other.begin(), other.end());
+	}
+	return *this;
+}
 
-	/// The value at the specified array index.
-	/// @warning Does not provide overflow protection.
-	///  @param[in]		i	The index of the value.
-	const int& operator[](int i) const { return m_data[i]; }
+template <typename T>
+class rcVector : public rcVectorBase<T, RC_ALLOC_TEMP> {
```

Comment:
Perhaps we should call this `rcTempVector` to make it more clear the situations for which it should be used?

Reply:
I figured "rcVector" is the common case in the code, and the few places where rcPermVector get used would get called out. Happy to change, though, if you think it would be confusing. 

Incidentally: Are you aware of users actually using different strategies for RC_ALLOC_*? It seems like it would be hard to do robustly without telling rcFree what type it is freeing.

Reply:
My main problem with it is that it is unclear that you should not use such a vector for long term storage. For example, when stored as a class member you should probably use `rcPermVector`.

I am not aware of users using different strategies, but a simple and practical one would be to use arena allocation for temporary memory, and regular heap memory for permanent memory. On free, it is easy to identify whether the memory is part of the arena or not (simple pointer compares). Then you just throw the entire arena away once your meshing process is completed.

Reply:
Sure, but if you don't know how big the arena is up front, you will likely need to grow it on demand, presumably with allocations from the heap, and then to find out what kind of pointer you have, you need to walk a table or something.

Reply:
Though, I guess you can just allocate an extra preamble with every rcAlloc, and use it to store info like this.

Reply:
It's been a while since the allocation hints were introduced. I remember vaguely something arena based as @jakobbotsch suggests. Most integrations I've done simply ignored the hint.

Usually it has been more important to differentiate between the memory used by Recast (memory used during building) and Detour (tile data from navmesh builder).

I'm not sure if it makes sense to have perm vector class. Perm memory size is known in advance, and you should not need vector for those (maybe there's case for verts/tris, not sure).

Reply:
There's several places where PERM memory is free'd and reallocated (see mergeContours) -- converting members to rcPermVector might help reduce some memory churn there. Additionally, using an rcPermVector instead of a naked pointer would give bounds-checking in debug builds for free.

Reply:
Here's relevant discussion from google code times:
https://code.google.com/archive/p/recastnavigation/issues/78
https://code.google.com/archive/p/recastnavigation/issues/110
https://code.google.com/archive/p/recastnavigation/issues/199

It looks like I'm happy panda. The commit history should have been imported to github, maybe there is a nugget or two in the commit messages from that time.

Reply:
@memononen I could see #307 solved by making `m_update` an `rcPermVector`. The alternative is to use a local `rcTempVector` in `dtTileCache::update`, but then we get allocation every time there is an update (except if we use small vector optimization). Perhaps the tiny dynamic allocation is ok since it would only happen when there were actual updates, however.